readthedocs/builds/models.py

```diff
@@ -705,3 +708,117 @@ def run_time(self):
         if self.start_time is not None and self.end_time is not None:
             diff = self.end_time - self.start_time
             return diff.seconds
+
+
+class VersionAutomationRule(PolymorphicModel, TimeStampedModel):
+
+    """Versions automation rules for projects."""
+
+    ACTIVATE_VERSION_ACTION = 'activate-version'
+    ACTIONS = (
+        (ACTIVATE_VERSION_ACTION, _('Activate version on match')),
+    )
+
+    project = models.ForeignKey(
+        Project,
+        related_name='automation_rules',
+        on_delete=models.CASCADE,
+    )
+    priority = models.IntegerField(
+        _('Rule priority'),
+        help_text=_('A lower number (0) means a higher priority'),
+    )
+    match_arg = models.CharField(
+        _('Match argument'),
+        help_text=_('Value used for the rule to match the version'),
+        max_length=255,
+    )
+    action = models.CharField(
+        _('Action'),
+        max_length=32,
+        choices=ACTIONS,
+    )
+    action_arg = models.CharField(
+        _('Action argument'),
+        help_text=_('Value used for the action to perfom an operation'),
+        max_length=255,
+        null=True,
+        blank=True,
+    )
+    version_type = models.CharField(
+        _('Version type'),
+        max_length=32,
+        choices=VERSION_TYPES,
+    )
+
+    class Meta:
+        unique_together = (('project', 'priority'),)
+        ordering = ('priority', '-modified', '-created')
+
+    def run(self, version, *args, **kwargs):
+        """
+        Run an action if `version` matches the rule.
+
+        :type version: readthedocs.builds.models.Version
+        :returns: True if the action was performed
+        """
+        if version.type == self.version_type:
+            match_result = self.match(version, self.match_arg)
+            if match_result is not None:
+                self.apply_action(version, match_result, self.action_arg)
+                return True
+        return False
+
+    def match(self, version, match_arg):
+        """
+        Returns an object different from None if the version matches the rule.
+
+        :type version: readthedocs.builds.models.Version
+        :param str match_arg: Additional argument to perform the match
+        :returns: An object different from `None` if the match is successful.
+                  The result will be passed to `apply_action`.
+        """
+        return None
+
+    def apply_action(self, version, match_result, action_arg):
```

Comment:
There is no need to pass `match_result` and `action_arg` as arguments. The object already have them under `self.`